File: scripts/e2e-vectors.mjs
/**
 * 生成 E2E 交叉验证向量（Node 侧权威实现 → Android Kotlin 单元测试）。
 *
 * 固定熵约定（Android 侧 E2eVectors 与测试必须按同一约定构造输入）：
 * - host 长期私钥 = RFC 7748 §5.2 Alice；设备长期私钥 = Bob；
 * - 配对 psk = 32 字节全 0x11；nonce_d = 32×0x22；nonce_h = 32×0x33；
 * - 连接握手临时私钥：设备 = 32×0x44，Host = 32×0x55；
 * - 测试房间路由：room=host-1, from=device-1, to=host-1。
 *
 * 用法：node scripts/e2e-vectors.mjs   （输出与 android E2eVectors.kt 对齐）
 */
import {
  deriveConfirmKey,
  derivePskRoot,
  deriveSessionKeys,
  handshakeMacFromDevice,
  handshakeMacFromHost,
  pairMacFromDevice,
  pairMacFromHost,
  privateKeyFromRaw,
  publicRawFromPrivateRaw,
  sealAead,
  toBase64Url,
} from "../packages/e2e/dist/index.js";
import { canonEnvelopeAad, envelopeNonce } from "../packages/protocol/dist/index.js";
import { createPublicKey, diffieHellman } from "node:crypto";

const SPKI_PREFIX = Buffer.from("302a300506032b656e032100", "hex");
const hostPriv = Buffer.from("77076d0a7318a57d3c16c17251b26645df4c2f87ebc0992ab177fba51db92c2a", "hex");
const devicePriv = Buffer.from("5dab087e624a8a4b79e17f8b83800ee66f3bb1292618b6fd1c2f8b27ff88e0eb", "hex");
const hostPub = publicRawFromPrivateRaw(hostPriv);
const devicePub = publicRawFromPrivateRaw(devicePriv);
const agreement = (privateRaw, peerPublicRaw) =>
  diffieHellman({
    privateKey: privateKeyFromRaw(privateRaw),
    publicKey: createPublicKey({ key: Buffer.concat([SPKI_PREFIX, peerPublicRaw]), format: "der", type: "spki" }),
  });

const ss = agreement(devicePriv, hostPub);
const pskRoot = derivePskRoot({ sharedSecret: ss, hostPublicRaw: hostPub, devicePublicRaw: devicePub });
const psk = Buffer.alloc(32, 0x11);
const confirmKey = deriveConfirmKey({ psk, pskRoot, hostPublicRaw: hostPub, devicePublicRaw: devicePub });
const nonceD = toBase64Url(Buffer.alloc(32, 0x22));
const nonceH = toBase64Url(Buffer.alloc(32, 0x33));

const eDPriv = Buffer.alloc(32, 0x44);
const eHPriv = Buffer.alloc(32, 0x55);
const ePubD = publicRawFromPrivateRaw(eDPriv);
const ePubH = publicRawFromPrivateRaw(eHPriv);
const ee = agreement(eDPriv, ePubH);
const keys = deriveSessionKeys({ sharedSecret: ee, pskRoot, hostEphemeralPublic: ePubH, deviceEphemeralPublic: ePubD });

const dataCt = (n, payload) =>
  toBase64Url(sealAead({
    key: keys.kDeviceToHost,
    nonce: envelopeNonce(n),
    aad: canonEnvelopeAad({ k: "data", room: "host-1", from: "device-1", to: "host-1", n }),
    plaintext: Buffer.from(payload, "utf8"),
  }));

console.log(`HOST_PUB    = ${toBase64Url(hostPub)}`);
console.log(`DEVICE_PUB  = ${toBase64Url(devicePub)}`);
console.log(`PSK_ROOT    = ${toBase64Url(pskRoot)}`);
console.log(`CONFIRM_KEY = ${toBase64Url(confirmKey)}`);
console.log(`MAC_D       = ${toBase64Url(pairMacFromDevice(confirmKey, nonceD))}`);
console.log(`MAC_H       = ${toBase64Url(pairMacFromHost(confirmKey, nonceD, nonceH))}`);
console.log(`E_PUB_D     = ${toBase64Url(ePubD)}`);
console.log(`E_PUB_H     = ${toBase64Url(ePubH)}`);
console.log(`K_H2D       = ${toBase64Url(keys.kHostToDevice)}`);
console.log(`K_D2H       = ${toBase64Url(keys.kDeviceToHost)}`);
console.log(`HS_MAC_H    = ${toBase64Url(handshakeMacFromHost(keys, ePubH, ePubD))}`);
console.log(`HS_MAC_D    = ${toBase64Url(handshakeMacFromDevice(keys, ePubH, ePubD))}`);
console.log(`DATA_CT_N1  = ${dataCt(1, '{"type":"device.ready"}')}`);
console.log(`DATA_CT_N2  = ${dataCt(2, "hello world")}`);

File: android/app/src/test/java/dev/pi/remote/E2eVectors.kt
package dev.pi.remote

/**
 * Node 侧权威实现（`packages/e2e`）用固定密钥生成的交叉验证向量。
 * 生成脚本：`scripts/e2e-vectors.mjs`。固定熵约定见该脚本头注释。
 * 两端任何一侧改动密码学行为，这些测试必挂。
 */
object E2eVectors {
    /** RFC 7748 §5.2 Alice 公钥（= host 长期公钥）。 */
    const val HOST_PUB = "hSDwCYkwp1R0i33ctD73Wg2_Og0mOBr066SpjqqbTmo"
    /** RFC 7748 §5.2 Bob 公钥（= 设备长期公钥）。 */
    const val DEVICE_PUB = "3p7bfXt9wbTTW2HC7OQ1Nz-DQ8hbeGdNrfx-FG-IK08"

    const val PSK_ROOT = "Xc5yaEsPIReHbncWPXU1GrW_YpUGNcpX5p1Zb-9dEwE"
    const val CONFIRM_KEY = "-G7rO7iQuhdH7rmm3agDYiIR8XmedYKDmimm7DAlHEE"

    /** 32 字节全 0x22 的 base64url。 */
    const val NONCE_D = "IiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiIiI"
    /** 32 字节全 0x33 的 base64url。 */
    const val NONCE_H = "MzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzMzM"

    const val MAC_D = "ywudy3_oqEafVQ1k7OtqWzUvrOG9ETuIwgiSAE0Em3c"
    const val MAC_H = "mhBLskBz0yVpG0VOkZBT9OPYN-VBFtIe8n5lLbbcBOw"

    /** 32 字节全 0x44 的 X25519 私钥对应公钥。 */
    const val E_PUB_D = "_y7kVgHsG2cxDHeQQEWFrmlzMe7hwfjPJBlzHB__Pms"
    /** 32 字节全 0x55 的 X25519 私钥对应公钥。 */
    const val E_PUB_H = "OKtmS9hvd9fma92a4HkpE6lP2LM6EmACfktGwfSITGc"

    const val K_H2D = "-Kwx43HwDDM6EXM8p80AX_mmYwMamhFMm089UZM1rWI"
    const val K_D2H = "a2vzVQwGvTKpwawGTEw6-qEO1HXnUC5P9uTSDTd6WgY"
    const val HS_MAC_H = "phYx4BZpBVq0X8qN-_qZ2tMZLobFaYILeNZfd0EDbVA"
    const val HS_MAC_D = "Lkw-uItTWBebh0ONoovU7ypS7-2EjpS3Lj5XmOk9TBM"

    /** data 帧密文：hdr = {k:data, room:host-1, from:device-1, to:host-1, ch:ctl, n:1}，payload = {"type":"device.ready"}。 */
    const val DATA_CT_N1 = "D9kX4K5xUPHFcmADHFlly_cs6SipR2RIN0tzKFh3Q3JA0mvUw7T8"
    /** 同上但 n:2，payload = "hello world"。 */
    const val DATA_CT_N2 = "Vnw5ZWj0XBmcrEZHu4ObE4FabcMXjWIESMkB"
}
